pyomo/contrib/pynumero/sparse/mpi_block_matrix.py

```diff
@@ -764,38 +764,145 @@ def __sub__(self, other):
-        rank = self._mpiw.Get_rank()
+    def _get_block_vector_for_dot_product(self, x):
+        if isinstance(x, MPIBlockVector):
+            """
+            Consider a non-empty block m_{i, j} from the mpi block matrix with rank owner r_m and the 
+            corresponding block v_{j} from the mpi block vector with rank owner r_v. There are 4 cases:
+              1. r_m = r_v
+                 In this case, all is good.
+              2. r_v = -1
+                 In this case, all is good.
+              3. r_m = -1 and r_v = 0
+                 All is good
+              4. If none of the above cases hold, then v_{j} must be broadcast
+            """
+            n_block_rows, n_block_cols = self.bshape
+            blocks_needing_broadcast = np.zeros(n_block_cols, dtype=np.int64)  # a value > 0 means broadcast
+            x_rank_ownership = x.rank_ownership
+            comm = self._mpiw
+            rank = comm.Get_rank()
+
+            if rank == 0:
+                block_indices = self._owned_mask
+            else:
+                block_indices = self._unique_owned_mask
+            block_indices = np.bitwise_and(block_indices, self._block_matrix._block_mask)
+            for i, j in zip(*np.nonzero(block_indices)):
+                r_m = self._rank_owner[i, j]
+                r_v = x_rank_ownership[j]
+                if r_m == r_v:
+                    pass
+                elif r_v == -1:
+                    pass
+                elif r_m == -1 and r_v == 0:
+                    pass
+                else:
+                    blocks_needing_broadcast[j] = 1
+
+            global_blocks_needing_broadcast = np.zeros(n_block_cols, dtype=np.int64)
+            comm.Allreduce(blocks_needing_broadcast, global_blocks_needing_broadcast)
+            indices_needing_broadcast = np.nonzero(global_blocks_needing_broadcast)[0]
+            if len(indices_needing_broadcast) == 0:
+                return x
+            else:
+                res = BlockVector(n_block_cols)
+                for ndx in np.nonzero(x.ownership_mask)[0]:
+                    res.set_block(ndx, x.get_block(ndx))
+                for j in indices_needing_broadcast:
+                    j_owner = x_rank_ownership[j]
+                    j_size = x.get_block_size(j)
+                    if rank == j_owner:
+                        data = x.get_block(j).flatten()
+                    else:
+                        data = np.empty(j_size)
+                    comm.Bcast(data, j_owner)
+                    res.set_block(j, data)
+                return res
+        elif isinstance(x, BlockVector):
+            return x
+        else:
+            assert isinstance(x, np.ndarray)
+            y = BlockVector(self.bshape[1])
+            for ndx, size in enumerate(self.col_block_sizes(copy=False)):
+                y.set_block(ndx, np.zeros(size))
+            y.copyfrom(x)
+            return y
+
+    def _block_vector_multiply(self, x):
```

Comment:
Did you intentionally delete the Parameters / return value from the docstring?

Reply:
Sort of. In the previous version, this method expected a `BlockVector`. Now, `x` can be a `BlockVector` or an `MPIBlockVector` as long as the conditions in the doctoring are satisfied. Since this is a private method, I was not worried too much about the doctoring.